contracts/BitGuildMarketplace.sol

```diff
@@ -0,0 +1,334 @@
+        require(msg.sender == address(PLAT), "Unauthorized PLAT contract address.");
+
+        address token;
+        uint tokenId;
+        (token, tokenId) = _decodeBuyData(_extraData);
+        bytes32 key = _getHashKey(token, tokenId);
+        address seller = listings[key].seller;
+
+        require(listings[key].currency == Currencies.PLAT, "Listing not in PLAT.");
+        require(_value > 0 && _value == listings[key].price, "Invalid price.");
+        require(listings[key].expiry > now, "Item expired.");
+
+        ERC721 gameToken = ERC721(token);
+        require(gameToken.ownerOf(tokenId) == address(this), "Item is not available.");
+
+        uint fee;
+        (,fee) = getFee(_value, _buyer, seller, token); // getFee returns percentFee and fee, we only need fee
+
+
+        // Transfer PLAT to marketplace contract
+        require(_PLAT.transferFrom(_buyer, address(this), _value), "PLAT payment transfer failed.");
+        // Transfer item token to buyer
+        gameToken.safeTransferFrom(this, _buyer, tokenId);
+        // Transfer Balance - fee to Seller
+        _PLAT.transfer(seller, _value - fee);
+
+        // delist item
+        _delist(key);
+
+        // Emit event
+        emit LogItemSold(_buyer, seller, token, tokenId, _value, listings[key].currency, now);
+    }
+
+    // ===========================================
+    // Admin Functions
+    // ===========================================
+    // @dev Update fee provider contract
+    function updateAllowWithdrawBeforeExpiry(bool _allow) public onlyOperator {
+        allowWithdrawBeforeExpiry = _allow;
+    }
+
+    // @dev Update fee provider contract
+    function updateFeeProvider(address _newAddr) public onlyOperator {
+        require(_newAddr != address(0), "Invalid contract address.");
+        FeeProvider = BitGuildFeeProvider(_newAddr);
+    }
+
+    // @dev Update whitelist contract
+    function updateWhitelist(address _newAddr) public onlyOperator {
+        require(_newAddr != address(0), "Invalid contract address.");
+        Whitelist = BitGuildWhitelist(_newAddr);
+    }
+
+    // @dev Update expiry date
+    function updateExpiry(uint _days) public onlyOperator {
+        require(_days > 0, "Invalid number of days.");
+        defaultExpiry = _days * 1 days;
+    }
+
+    // @dev Admin function: withdraw ETH balance
+    function withdrawETH() public onlyOwner payable {
+        address(this).transfer(msg.value);
+    }
+
+    // @dev Admin function: withdraw PLAT balance
+    function withdrawPLAT() public onlyOwner payable {
+        uint balance = PLAT.balanceOf(this);
+        PLAT.transfer(msg.sender, balance);
+    }
+
+    // ===========================================
+    // Internal Functions
+    // ===========================================
+    // @dev clear listing for hash key
+    function _delist(bytes32 _key) internal {
+        // remove listing
+        listings[_key].currency = Currencies(0);
```

Comment:
I think there should be a single method to delete a mapping entry. This is fragile, imagine you add a new param and forget to update here.